File: ai-solutions/03/gpt-5.1-codex/js/solution.js
const fs = require("fs");

function maxSubsequenceNumber(line, k) {
  const digits = line.trim();
  if (k >= digits.length) return BigInt(digits);

  let remove = digits.length - k;
  const stack = [];

  for (const ch of digits) {
    while (remove > 0 && stack.length > 0 && stack[stack.length - 1] < ch) {
      stack.pop();
      remove--;
    }
    stack.push(ch);
  }

  const kept = stack.slice(0, k).join("");
  return BigInt(kept);
}

function main() {
  if (process.argv.length < 3) {
    console.error("Usage: node solution.js <input_file>");
    process.exit(1);
  }

  const inputPath = process.argv[2];
  const lines = fs.readFileSync(inputPath, "utf8").trim().split("\n");

  let part1 = 0n;
  let part2 = 0n;

  for (const line of lines) {
    if (!line.trim()) continue;
    part1 += maxSubsequenceNumber(line, 2);
    part2 += maxSubsequenceNumber(line, 12);
  }

  console.log(part1.toString());
  console.log(part2.toString());
}

main();

File: ai-solutions/03/gpt-5.1-codex/python/solution.py
import sys
from typing import List


def max_subsequence_number(line: str, k: int) -> int:
    """
    Return the largest possible number (as int) by keeping k digits
    from the given digit string while preserving order.
    """
    line = line.strip()
    if k >= len(line):
        return int(line)

    remove = len(line) - k
    stack: List[str] = []

    for ch in line:
        while remove and stack and stack[-1] < ch:
            stack.pop()
            remove -= 1
        stack.append(ch)

    # If we still have extra digits, trim from the right.
    stack = stack[:k]
    return int("".join(stack))


def solve(lines: List[str]) -> None:
    part1 = 0
    part2 = 0

    for line in lines:
        line = line.strip()
        if not line:
            continue
        part1 += max_subsequence_number(line, 2)
        part2 += max_subsequence_number(line, 12)

    print(part1)
    print(part2)


def main() -> None:
    if len(sys.argv) < 2:
        print("Usage: python solution.py <input_file>")
        sys.exit(1)

    input_path = sys.argv[1]
    with open(input_path, "r", encoding="utf-8") as f:
        lines = f.readlines()

    solve(lines)


if __name__ == "__main__":
    main()
